consensus/qsc/messages.go

```diff
@@ -0,0 +1,181 @@
+package qsc
+
+import (
+	"go.dedis.ch/dela/consensus/qsc/json"
+	"go.dedis.ch/dela/serde"
+	"golang.org/x/xerrors"
+)
+
+// Proposal is a message to wrap a proposal so that it can be deserialized
+// alongside other request messages.
+//
+// - implements serde.Message
+type Proposal struct {
+	serde.UnimplementedMessage
+
+	value serde.Message
+}
+
+// VisitJSON implements serde.Message. It serializes the proposal message in
+// JSON format.
+func (p Proposal) VisitJSON(ser serde.Serializer) (interface{}, error) {
+	value, err := ser.Serialize(p.value)
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't serialize value: %v", err)
+	}
+
+	m := json.Proposal{
+		Value: value,
+	}
+
+	return json.Request{Proposal: &m}, nil
+}
+
+// Message is a message sent by a node for the round.
+//
+// - implements serde.Message
+type Message struct {
+	serde.UnimplementedMessage
+
+	node  int64
+	value serde.Message
+}
+
+// VisitJSON implements serde.Message.
+func (m Message) VisitJSON(ser serde.Serializer) (interface{}, error) {
+	var data []byte
+	var err error
+
+	switch value := m.value.(type) {
+	case MessageSet:
+		data, err = ser.Serialize(value)
+	default:
+		data, err = ser.Serialize(Proposal{value: value})
+	}
+
+	if err != nil {
+		return nil, xerrors.Errorf("couldn't serialize value: %v", err)
+	}
+
+	msg := json.Message{
+		Node:  m.node,
+		Value: data,
+	}
+
+	return msg, nil
+}
+
+// MessageSet is a set of messages mapped by node index.
+//
+// - implements serde.Message
+type MessageSet struct {
+	serde.UnimplementedMessage
+
+	messages map[int64]Message
+	timeStep uint64
+	node     int64
+}
+
+// VisitJSON implements serde.Message.
+func (mset MessageSet) VisitJSON(ser serde.Serializer) (interface{}, error) {
+	messages := make(map[int64]json.Message)
+	for index, msg := range mset.messages {
+		raw, err := msg.VisitJSON(ser)
+		if err != nil {
+			return nil, xerrors.Errorf("couldn't serialize message: %v", err)
+		}
+
+		messages[index] = raw.(json.Message)
```

Comment:
this always makes me anxious, but yes, there is a priori no reason that `msg.VisiJSON` returns another type

Reply:
Yes, it will panic in unit tests if someone changes it.